cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesStore.java

```diff
@@ -133,79 +149,71 @@ private void deleteConfig(String namespace, String id) {
   }
 
   public void setProperties(Map<String, String> propMap) throws NotFoundException, ProfileConflictException {
-    setConfig(EMPTY_NAMESPACE, getMultipartKey(INSTANCE_PROPERTIES), propMap);
+    setConfig(EMPTY_NAMESPACE, new InstanceId(EMPTY_NAMESPACE), propMap);
   }
 
   public void setProperties(String namespace,
                             Map<String, String> propMap) throws NotFoundException, ProfileConflictException {
-    setConfig(namespace, getMultipartKey(namespace), propMap);
+    setConfig(namespace, new NamespaceId(namespace), propMap);
   }
 
   public void setProperties(String namespace, String appId, Map<String, String> propMap)
     throws NotFoundException, ProfileConflictException {
-    setConfig(namespace, getMultipartKey(namespace, appId), propMap);
+    setConfig(namespace, new ApplicationId(namespace, appId), propMap);
   }
 
   public void setProperties(String namespace, String appId, String programType, String programId,
                             Map<String, String> propMap) throws NotFoundException, ProfileConflictException {
-    setConfig(namespace, getMultipartKey(namespace, appId, programType, programId), propMap);
+    setConfig(namespace, new ProgramId(namespace, appId, ProgramType.valueOfCategoryName(programType),
+                                       programId), propMap);
   }
 
   public void deleteProperties() {
-    deleteConfig(EMPTY_NAMESPACE, getMultipartKey(INSTANCE_PROPERTIES));
+    deleteConfig(EMPTY_NAMESPACE, new InstanceId(EMPTY_NAMESPACE));
   }
 
   public void deleteProperties(String namespace) {
-    deleteConfig(namespace, getMultipartKey(namespace));
+    deleteConfig(namespace, new NamespaceId(namespace));
   }
 
   public void deleteProperties(String namespace, String appId) {
-    deleteConfig(namespace, getMultipartKey(namespace, appId));
+    deleteConfig(namespace, new ApplicationId(namespace, appId));
   }
 
   public void deleteProperties(String namespace, String appId, String programType, String programId) {
-    deleteConfig(namespace, getMultipartKey(namespace, appId, programType, programId));
-  }
-
-  private String getMultipartKey(String... parts) {
```

Comment:
Because I feel like it is very hard to use the config dataset, using entity id seems more straightforward and convenient. If we want to use this as a key, we have to make it public or copy it somewhere to generate the key

Reply:
It's fine if you change the method signatures to use EntityId, as long as the existing calls don't break. But the key format should not be public, it must be internal to the class that is doing the writing. If it needs to be public, that means there is some abstraction that is wrong. Others shouldn't have to know how this class is storing its data.

Reply:
So if we want to use the ConfigDataset directly and we want to retrieve a preference, we will have to generate the key for it. I am not sure if what is the best way to generate a way for it. And the way we store config is a bit confused to me. For dashboard, it just generates a random UUID for it, which basically means the get() method is completely useless for dashboard

Reply:
Moved the logic to PreferencesDataset